tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,479 @@
+
+```haskell
+deleteTodo :: Connection
+        -> Int
+        -> IO ()
+deleteTodo conn todoId = do
+  let query = [i| DELETE FROM todos
+                  WHERE id = ?|]
+
+  quickQuery' conn query [toSql todoId]
+  commit conn
+```
+
+Finally, `allTodos` selects all the Todo items while organizing them with the completed
+ones listed last. Bear in mind `quickQuery'` returns a `[[SqlValue]]` value in the
+context of an [IO monad](https://wiki.haskell.org/IO_inside), so it uses a simple
+and easy to understand [`rowToTodo`](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Db.hs#L80-L87) function to turn those rows into `Todo`
+data values.
+
+```haskell
+allTodos  :: Connection
+          -> IO [Todo]
+allTodos conn = do
+  let query = [i| SELECT id, description, completed
+                  FROM todos
+                  ORDER BY completed ASC|]
+  map rowToTodo <$> quickQuery' conn query []
+
+```
+
+## Now you can see
+
+Great, all is ready now to serve some content. This is a simple application so
+it has a unique view which consists in a Todo items list with a couple of buttons
+to mark it as done or delete it, and a form for new items submission purpose. Just
+like this
+
+<p align="center">
+  <img alt="Snap Todo web app" src="todo.png" style="width:50%;height:50%;" />
+  </br>
+</p>
+
+
+This application uses Snap and therefore it seemed logical to use its default template engine: Heist.
+
+Heist uses simple XML tags to bind values to your templates in a straightforward
+way. It's a great library which even allows to use compiled or interpreted templates,
+but that's out of the scope of the tutorial. Anyway, you should know this
+tutorial uses interpreted templates, and also another structure called [Splice](http://snapframework.com/docs/tutorials/heist#heist-programming) to pull
+values from. You'll see more about that in a bit
+
+This is an extract of how the [todos.tpl](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/snaplets/heist/templates/todos.tpl) Heist template for this tutorial
+application looks like. The submission form was left out for the sake of simplicity
+
+```html
+<apply template="base">
+  <div class="col-xs-6 col-xs-offset-3">
+    <div class="well">
+      <h1>TODO Items</h1>
+      <ul class="list-group">
+        <todos>
+          <input type="hidden" value="${id}">
+          <li class="list-group-item clearfix ${successClass}">
+            <span><description/></span>
+            <div class="pull-right" role="group">
+              <a class="btn btn-xs btn-success img-circle"
+                  onclick="document.getElementById('form-check-${id}').submit();">
+                  ✓
+              </a>
+              <form id="form-check-${id}" action="/todos/${id}/complete/${notCompleted}" method="post" class="hidden-form">
+              </form>
+              <span> </span>
+              <a class="btn btn-xs btn-danger img-circle"
```

Comment:
This Javascript seems wholly unnecessary to me. Make the x and check individual forms each.